File: programs/3-benchmarks+codegen/valid/_mergesort.py
def merge(array, temp, lo, hi):
    mid = (lo+hi)/2
    temp_index = lo
    index_left = lo
    index_right = mid+1

    while index_left <= mid and index_right <= hi:
        if array[index_left] < array[index_right]:
            temp[temp_index] = array[index_left]
            index_left += 1
        else:
            temp[temp_index] = array[index_right]
            index_right += 1

        temp_index += 1

    while index_right <= hi:
        temp[temp_index] = array[index_right]
        index_right += 1
        temp_index += 1

    while index_left <= mid:
        temp[temp_index] = array[index_left]
        index_left += 1
        temp_index += 1

    for i in xrange(lo, hi+1):
        array[i] = temp[i]

def merge_sort_helper(array, temp, lo, hi):
    if lo >= hi:
        return

    mid = (lo+hi)/2
    merge_sort_helper(array, temp, lo, mid)
    merge_sort_helper(array, temp, mid+1, hi)

    merge(array, temp, lo, hi)

def merge_sort(array, length):
    temp = []
    for i in xrange(0, length):
        temp.append(array[i])
    merge_sort_helper(array, temp, 0, length-1)

def main():
    array = []
    length = 10000000
    for i in xrange(0, length):
        array.append(length-i-1)

    merge_sort(array, length)

    for i in range(10):
        print array[i]

if __name__  == "__main__":
    main()

File: programs/3-benchmarks+codegen/valid/mergesort.go
//~0
//~1
//~2
//~3
//~4
//~5
//~6
//~7
//~8
//~9
//&Python: 1 minutes 8.9 seconds
//&C++: 6.9 seconds

package main

func merge(array, temp []int, lo, hi int) {
    mid := (lo+hi)/2
    temp_index := lo
    index_left := lo
    index_right := mid+1

    for index_left <= mid && index_right <= hi {
        if (array[index_left] < array[index_right]) {
            temp[temp_index] = array[index_left]
            index_left++
        } else {
            temp[temp_index] = array[index_right]
            index_right++
        }

        temp_index++
    }

    for index_right <= hi {
        temp[temp_index] = array[index_right]
        index_right++
        temp_index++
    }

    for index_left <= mid {
        temp[temp_index] = array[index_left]
        index_left++
        temp_index++
    }

    for i := lo; i <= hi; i++ {
        array[i] = temp[i]
    }

}

func merge_sort_helper(array, temp []int, lo, hi int) {
    if (lo >= hi) {
        return
    }

    mid := (lo+hi)/2
    merge_sort_helper(array, temp, lo, mid)
    merge_sort_helper(array, temp, mid+1, hi)

    merge(array, temp, lo, hi)
}

func merge_sort(array []int, length int) {
    var temp []int
    for i := 0; i < length; i++ {
        temp = append(temp, array[i])
    }
    merge_sort_helper(array, temp, 0, length-1)
}

func main() {
    // Create the array to sort
    var array []int
    length := 10000000
    for i := 0; i < length; i++ {
        array = append(array, length-i-1);
    }

    // Sort the array
    merge_sort(array, length)

    // Print the sorted array
    for i := 0; i < 10; i++ {
        println(array[i])
    }
}
